clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -253,35 +254,27 @@ public synchronized FutureTransactionalResult initializeTransactions() {
 
     public synchronized void beginTransaction() {
         ensureTransactional();
+        maybeFailWithError();
         if (!transitionTo(State.IN_TRANSACTION))
-             throw new IllegalStateException("Producer isn't ready to begin a transaction.");
+             throw new KafkaException("Producer isn't ready to begin a transaction, most likely because there is " +
+                     "already an ongoing transaction.");
     }
 
     public synchronized FutureTransactionalResult beginCommittingTransaction() {
         ensureTransactional();
-        if (!transitionTo(State.COMMITTING_TRANSACTION)) {
-            String msg = "Cannot commit transaction, either because a transaction is already " +
-                    "being completed at the moment, or because there has been an error with a previous request.";
-            if (lastError != null)
-                throw new IllegalStateException(msg, lastError);
-            else
-                throw new IllegalStateException(msg);
-        }
-
+        maybeFailWithError();
+        if (!transitionTo(State.COMMITTING_TRANSACTION))
+            throw new KafkaException("Cannot commit transaction, most likely because a transaction is already being completed.");
         return beginCompletingTransaction(true);
     }
 
     public synchronized FutureTransactionalResult beginAbortingTransaction() {
         ensureTransactional();
-        if (!transitionTo(State.ABORTING_TRANSACTION)) {
-            String msg = "Cannot abort transaction, either because a transaction is already " +
-                    "being completed at the moment, or because there has been an error with a previous request.";
-            if (lastError != null)
-                throw new IllegalStateException(msg, lastError);
-            else
-                throw new IllegalStateException(msg);
-        }
-
+        if (isFenced())
+            throw new ProducerFencedException("There is a newer producer using the same transactional.id.");
+        if (!transitionTo(State.ABORTING_TRANSACTION))
```

Comment:
Never mind. As was explained offline, the last error is reset once the transaction is completed.